core/templates/components/button-directives/primary-button.component.ts

```diff
@@ -13,49 +13,94 @@
 // limitations under the License.
 
 /**
- * @fileoverview Component for the primary buttons displayed on static pages.
+ * @fileoverview Component for primary buttons displayed on static pages.
  */
 
-import { Component, Input, Output, EventEmitter } from '@angular/core';
+import { Component, Input, Output, EventEmitter, OnInit } from '@angular/core';
 import { downgradeComponent } from '@angular/upgrade/static';
-import { WindowRef } from
-  'services/contextual/window-ref.service';
 import './primary-button.component.css';
+import { WindowRef } from 'services/contextual/window-ref.service';
 
 
 @Component({
   selector: 'oppia-primary-button',
   templateUrl: './primary-button.component.html',
   styleUrls: ['./primary-button.component.css']
 })
-export class PrimaryButtonComponent {
+export class PrimaryButtonComponent implements OnInit {
+  /**
+   * The main text to be displayed on the button.
+   */
   @Input() buttonText: string = '';
+
+  /**
+   * An optional array of custom CSS classes to be applied to the component.
+   */
   @Input() customClasses?: string[];
+
+  /**
+   * The URL the button should navigate to when clicked, optional.
+   * If no buttonHref is provided then the component acts like an HTML button
+   * element instead of an anchor link element.
+   */
+  @Input() buttonHref: string = '#';
+
+  /**
+   * A boolean value indicating whether the button should be disabled, optional.
+   * Valid only when the component is an HTML button element and not an anchor.
+   */
   @Input() disabled?: boolean = false;
-  @Input() buttonHref: string | null = null;
+
+  /**
+   * The function to execute when the component is clicked, optional.
+   * The return type of the function must be void.
+   */
   @Output() onClickPrimaryButton: EventEmitter<void> = new EventEmitter<void>();
 
+  componentIsButton: boolean = false;
+  openInNewTab: boolean = false;
+
   constructor(
     private windowRef: WindowRef,
   ) {}
 
-  handleButtonClick(): void {
+  ngOnInit(): void {
+    this.componentIsButton = this.buttonHref === '#';
+  }
+
+  getButtonHref(): string {
+    return this.buttonHref;
+  }
+
+  getTarget(): string {
+    this.openInNewTab = this.isExternalLink(this.buttonHref);
+    return this.openInNewTab ? '_blank' : '_self';
+  }
+
+  private isExternalLink(link: string): boolean {
+    return link.startsWith('http://') || link.startsWith('https://');
+  }
+
+  handleButtonClick(event: MouseEvent): void {
+    event.preventDefault();
+
     if (this.onClickPrimaryButton.observers.length > 0) {
       this.onClickPrimaryButton.emit();
```

Comment:
There is a section [here](https://developers.google.com/analytics/devguides/collection/gtagjs/sending-data#handle_timeouts) explaining how to handle the case when the gtag.js library fails to load. Do we need to handle it here (e.g. catch any thrown exceptions, use a timeout, etc.)?

Also is it worth adding tests for these cases?

Reply:
I think that's redundant for our use-cases, as adding a timeout would just delay executing the subsequent functions and thus delay loading the next page more. They aren't adding a timeout or anything to sending the event to google analytics, but they have an `event_callback` where the timeout is relevant - which we don't. Does that make sense?

Reply:
Yes, but it doesn't quite get at the main point of my question -- sorry. 

Let me put this another way. Suppose gtag.js fails to load. What happens then? Should we add tests for this case, too?

Reply:
Sorry for misunderstanding, in case a function passed to the component like registering analytics doesn't work, it would just carry on with what it's supposed to do aside from throwing errors in the console.

Reply:
@Vir-8 But how did you test this? I ask because emit() is synchronous (I think) so sometimes what can happen in cases like this is that the function throws an error and stops the execution at the point of the error. This is why I'm trying to check that this doesn't happen here. Can you show a video?

Reply:
Tested it by passing invalid functions and misspelling things beforehand, like having gtagg instead of gtag in the function that sends event to Google analytics.

The emit function is indeed synchronous but a function emitted by the event emitter failing doesn't cause the main function to stop executing, it carries on unless told to do otherwise. Here's some examples -

1. Changing `this.siteAnalyticsService....` to `this.siteAnalyticssService....`
[Screencast from 20-02-24 06:15:48 PM IST.webm](https://github.com/oppia/oppia/assets/73329256/35884e31-0c69-4894-9f84-d76ad6582f5d)

2. Changing an instance of gtag to gtagg in site-analytics
[Screencast from 20-02-24 06:19:36 PM IST.webm](https://github.com/oppia/oppia/assets/73329256/6c99317d-b72a-4493-b521-937151f7c0cf)

Reply:
OK, thanks for confirming! LGTM.